packages/devtools/src/flows/config.load.ts

```diff
@@ -0,0 +1,111 @@
+ */
+export const configLoadFlow = async <TOmniGraph = OmniGraph>({
+    configPath,
+    configSchema,
+    logger = createLogger(),
+}: ConfigLoadFlowArgs<TOmniGraph>): Promise<TOmniGraph> => {
+    logger.verbose(`Loading config from ${configPath}`)
+
+    const absolutePath = resolve(configPath)
+    logger.verbose(`Resolved config file location for '${configPath}': '${absolutePath}'`)
+
+    // First we check that the config file is indeed there and we can read it
+    logger.verbose(`Checking config file '${absolutePath}' for existence & readability`)
+    const isConfigReadable = isFile(absolutePath) && isReadable(absolutePath)
+    if (!isConfigReadable) {
+        throw new Error(
+            `Unable to read config file '${configPath}'. Check that the file exists and is readable to your terminal user`
+        )
+    }
+
+    // Keep talking to the user
+    logger.verbose(`Config file '${absolutePath}' exists & is readable`)
+
+    // Now let's see if we can load the config file
+    let rawConfig: unknown
+    try {
+        logger.verbose(`Loading config file '${absolutePath}'`)
+
+        rawConfig = await importDefault(absolutePath)
+    } catch (error) {
+        throw new Error(`Unable to read config file '${configPath}': ${error}`)
+    }
+
+    logger.verbose(`Loaded config file '${absolutePath}'`)
+
+    // Now let's check whether the config file contains a function
+    //
+    // If so, we'll execute this function and will expect a config as a result
+    let rawConfigMaterialized: unknown
+    if (typeof rawConfig === 'function') {
+        logger.verbose(`Executing configuration function from config file '${absolutePath}'`)
+
+        try {
+            rawConfigMaterialized = await rawConfig()
+        } catch (error) {
+            throw new Error(`Got an exception while executing config funtion from file '${configPath}': ${error}`)
+        }
+    } else {
+        logger.verbose(`Using exported value from config file '${absolutePath}'`)
+        rawConfigMaterialized = rawConfig
+    }
+
+    // It's time to make sure that the config is not malformed
+    //
+    // At this stage we are only interested in the shape of the data,
+    // we are not checking whether the information makes sense (e.g.
+    // whether there are no missing nodes etc)
+    logger.verbose(`Validating the structure of config file '${absolutePath}'`)
+
+    // We'll try/catch the schema validation (even though we are using the "safe" version,
+    // zod will just throw if any of the schema transformations throw)
+    //
+    // We do this so that we can prepend the error message with a more meaningful one
+    let configParseResult: SafeParseReturnType<unknown, TOmniGraph>
+    try {
+        configParseResult = await configSchema.safeParseAsync(rawConfigMaterialized)
+    } catch (error) {
+        throw new Error(`Config from file '${configPath}' is invalid: ${error}`)
+    }
+
+    if (configParseResult.success === false) {
```

Comment:
Should we write this as `!configParseResult.success` if the value can ever be `null`, `undefined`, or other falsy values?

Reply:
Oh in this case it's always `true` or `false`. And TypeScript also has this quirk where `!` actually breaks (type narrowing)[https://www.typescriptlang.org/docs/handbook/2/narrowing.html].

```typescript
const a: boolean = false;

// If we check for true/false directly, type narrowing works 
if (a === false) return

// i.e. a will by typed as `true` now
const typeNarrowedA: true = a

// BUT if instead we check with `!`, type narrowing stops working
if (!a) return

// And this statement will not compile
const typeNarrowedA: true = a
```